<launch>
  <arg name="namespace"                 default="iris"/>
  <arg name="world_name"                default="basic"/> <!-- Name of your Gazebo world file (e.g., basic.world) -->
  <arg name="paused"                    default="false"/>
  <arg name="gui"                       default="true"/>
  <arg name="debug"                     default="false"/>
  <arg name="verbose"                   default="false"/>
  <arg name="use_sim_time"              default="true"/> <!-- Essential for Gazebo -->


  <include file="$(find fly_bot)/launch/spawn_rotors_iris.launch">
    <arg name="namespace" value="$(arg namespace)"/>
    <arg name="world_name" value="$(arg world_name)"/>
    <arg name="paused"     value="$(arg paused)"/>
    <arg name="gui"        value="$(arg gui)"/>
    <arg name="debug"      value="$(arg debug)"/>
    <arg name="verbose"    value="$(arg verbose)"/>
    <arg name="x"          value="20.0"/>
    <arg name="y"          value="0.0"/>
    <arg name="z"          value="0.1"/>
    <arg name="yaw"        value="0.0"/>
  </include>

  <!-- 3. Launch the Controller Node -->
  <!-- Note: Assumes your script is now named lemniscate_c3bf.py -->
  <node pkg="fly_bot" type="lemniscate_c3bf.py"
        name="clf_iris_trajectory_controller" output="screen">

    <!-- Parameters loaded here -->
    <param name="namespace" value="$(arg namespace)" /> <!-- Pass namespace -->
    <param name="use_model_states" value="true" /> <!-- Use Gazebo ModelStates -->


    <!-- Drone Physical Properties -->
    <param name="mass" value="1.5" />
    <param name="gravity" value="9.81" />
    <param name="I_x" value="0.0348" />
    <param name="I_y" value="0.0459" />
    <param name="I_z" value="0.0977" />
    <param name="motor_constant" value="8.54858e-06" />
    <param name="moment_constant" value="1.3677728e-07" />
    <param name="max_rot_velocity" value="838.0" />
    <param name="min_thrust_factor" value="0.1" />
    <param name="gravity_comp_factor" value="1.022" />
    <param name="max_tilt_angle_deg" value="30.0" /> <!-- Reduced tilt for trajectory test -->

    <!-- Helix Trajectory Parameters (or Lemniscate if script changed) -->
    <!-- Keep these relevant to the trajectory your script generates -->
    <param name="static_obstacles"
       value="[ [10.00,  0.00, 3.00, 1.00],
                [ 3.09,  2.94, 3.00, 1.00],
                [-8.09, -4.76, 3.00, 1.00],
                [-6.47,  3.80, 3.00, 0.50],
                [ 3.71, -1.53, 3.00, 1.00] ]" />


    <param name="helix_start_diameter" value="40.0" />
    <param name="helix_end_diameter" value="15.0" />
    <param name="helix_height" value="30.0" />
    <param name="helix_laps" value="4.0" />
    <param name="trajectory_omega" value="0.2" />
    <param name="fixed_yaw_deg" value="0.0" />

    <!-- Takeoff/Hover Parameters -->
    <param name="takeoff_x" value="20.0" />
    <param name="takeoff_y" value="0.0" />
    <param name="takeoff_height" value="3.0" />
    <param name="hover_pos_threshold" value="0.15" />
    <param name="hover_vel_threshold" value="0.1" />
    <param name="hover_stabilization_secs" value="2.0" />

    <!-- Controller Gains -->
    <!-- Takeoff Gains -->
    <param name="k_takepos1" value="0.22" />
    <param name="k_takepos2" value="0.8" />
    <param name="k_takeatt1" value="2.05" />
    <param name="k_takeatt2" value="4.1" />
    <!-- Trajectory Gains -->
    <param name="k_trajpos1" value="0.8" />
    <param name="k_trajpos2" value="4.5" />
    <param name="k_trajatt1" value="16.0" />
    <param name="k_trajatt2" value="32.0" />

    <!-- CBF Parameters (Still need defaults even if no obstacles) -->
    <param name="cbf_gamma" value="1.0"/>
    <param name="drone_radius" value="0.5" />

    <!-- Control Loop Rate -->
    <param name="control_rate" value="100.0" />


  </node>


  <node name="trajectory_plotter_node"
        pkg="fly_bot"
        type="lemniscate_c3bf_plotter.py" 
        output="screen">
    <!-- Provide necessary topics -->
    <param name="odom_topic"   value="/iris/ground_truth/odometry"/>
    <param name="state_topic"  value="/clf_iris_trajectory_controller/control/state"/>
    <!-- Provide obstacles LIST AGAIN for the plotter -->
    <param name="static_obstacles"
       value="[ [10.00,  0.00, 3.00, 1.00],
                [ 3.09,  2.94, 3.00, 1.00],
                [-8.09, -4.76, 3.00, 1.00],
                [-6.47,  3.80, 3.00, 0.50],
                [ 3.71, -1.53, 3.00, 1.00] ]" />

    <!-- Other plotter params like save_dir, helix params (if needed for desired path) -->
    <param name="plot_save_dir"        value="$(find fly_bot)/plots"/>
    <param name="helix_start_diameter" value="40.0" />
    <param name="helix_end_diameter"   value="15.0" />
    <param name="helix_height"         value="30.0" />
    <param name="helix_laps"           value="4.0" />
    <param name="trajectory_omega"     value="0.07" />
  </node>
</launch>